### ISSUE TITLE
Implement an approval Workflow

### ISSUE THREAD

**Issue by dasistwas** [Wunderbyte-GmbH/moodle-mod_datalynx#160]
In order to publish a course, a approval workflow should be implemented, that generates E-Mails to certain persons or roles. The process is cascading: 

1) Teacher submits the datalynx instance with all the entries so that the study director will get an e-mail. 
2) The study director has a look at the datalynx instance and if everything ist correct
3) The study director approves it. Mails goes to the office so they can take the next steps. The mail has to include the course link or the course id the datalynx instance is used in.

In general: teachers are creating the semester plan for their students with a datalynx instance in their course. The teacher adds one entry for each study phase. The final datalynx instance is composed of several entries with workload, and description of the phase. The workload is calculated (JS implementation) for all the entries together. When the plan is approved by the study director, the office staff will enrol the students in the course.

### Possible implementation

There is already a status field in datalynx, which could be used.  But the feature wanted is more based on the whole instance than on a single entry. So there could be a separate view for approval. So maybe that can be created as rule or event type notification with conditions.

**Comment by michaelpollak** [comment]
![image](https://user-images.githubusercontent.com/475798/121890866-18ded580-cd1b-11eb-803b-4c09061cdb7c.png)

Extended the UI to allow a rule that only triggers on specific checkboxes and sends mail to specific user. Currently testing and working on an issue where mail is sent everytime an entry is updated and the checkbox is marked. Will update after I've found a neat fix.

**Comment by dasistwas** [comment]
great. just to be sure: "checkbox" is the name of the checkbox and not the "type" of the element?

**Comment by michaelpollak** [comment]
Yes exactly, I named the field of type checkbox "checkbox" :)
For the UI we are looking for all checkboxes in this datalynx instance and offer a dropdown list for selection. The default is "All" to keep backwards compatibility. For easier usage the area "Trigger only on specific element" is hidden until the Trigger event "Entry updated" is selected.

**Comment by michaelpollak** [comment]
Workflow for approval:

- Set up a field checkbox "waiting for approval"
- Set up a field checkbox "Approved"
- Set up a view "Approval workflow"
- Add the two checkboxes
- Add Event notification "Approved" that triggers on updates of checkbox "Approved" and sends mail to "Other user"
- Add Event notification "Waiting for approval" that triggers on updates of checkbox "Waiting for approval" and sends mail to "Other user"

This whole setup can be reused for every course. When checkbox "Waiting" is marked by teacher mail is sent out.
When checkbox "Approved" is marked and "Waiting for approval" is removed mails are sent out.

Looking forward to your feedback and improvements, thank you, have a nice day.

**Comment by dasistwas** [comment]
@michaelpollak we would need the trigger also on "entry created". Maybe limiting to entry updated is not necessary at all...

**Comment by dasistwas** [comment]
When I assign a manager to the course category and then trigger the event and choose as the receiver the role manager I get that error:

```
Sending message to the user with id 11
++ Attempt to send msg from a provider mod_datalynx/event_entry_created that is inactive or not allowed for the user id=11 ++
* line 224 of /lib/messagelib.php: call to debugging()
* line 46 of /mod/datalynx/classes/task/sendmessage_task.php: call to message_send()
* line 333 of /lib/cronlib.php: call to mod_datalynx\task\sendmessage_task->execute()
* line 198 of /lib/cronlib.php: call to cron_run_inner_adhoc_task()
* line 76 of /lib/cronlib.php: call to cron_run_adhoc_tasks()
* line 79 of /admin/cli/cron.php: call to cron_run()
Sent.
```

**Comment by michaelpollak** [comment]
Yes it is built on moodle events so the usual permissions apply. 
By default we only allow teachers and non-editing-teachers to receive "notifyentryupdate", we could change this default to integrate the managers as well if you like to? Or just have the admins decide who can receive event updates and add it somewhere as a note in the user interface?
